### ISSUE TITLE
DefaultRecordUnmapper.IterableUnmapper uses name based field access instead of index based ones, leading to ambiguities

### ISSUE THREAD

**Issue by sgrimm** [jOOQ/jOOQ#12704]
### Expected behavior

Columns in a multiset should behave like top-level columns: if you include a column twice in the `SELECT` clause, the same value should appear twice in the row (assuming it's just a plain column, not a function or a sequence or so forth).

### Actual behavior

With a single-level multiset, the value of the first occurrence of the column is correctly populated, but the second occurrence has a null value.

With a two-or-more-level multiset, the real value is completely discarded and the record has two nulls.

### Steps to reproduce the problem

```sql
CREATE TABLE projects (id INTEGER PRIMARY KEY);
INSERT INTO projects VALUES (1);
```

```kotlin
  val nestedResults =
      dslContext
          .select(DSL.multiset(DSL.select(PROJECTS.ID, PROJECTS.ID).from(PROJECTS)))
          .fetch()
  println(nestedResults)

  val doubleNestedResults =
      dslContext
          .select(
              DSL.multiset(
                  DSL.select(
                      DSL.multiset(DSL.select(PROJECTS.ID, PROJECTS.ID).from(PROJECTS)))))
          .fetch()
  println(doubleNestedResults)
```

Output:

```
+-------------+
|multiset     |
+-------------+
|[(1, {null})]|
+-------------+

+----------------------+
|multiset              |
+----------------------+
|[([({null}, {null})])]|
+----------------------+
```

The workaround, of course, is to stop including the same column twice. That can be tricky if the query is dynamically constructed.

### Versions

- jOOQ: 3.15.4
- Java: OpenJDK 15.0.1
- Database (include vendor): PostgreSQL 13.4
- OS: MacOS 12.0.1
- JDBC Driver (include name if inofficial driver): 42.2.24

**Comment by lukaseder** [comment]
Yeah, huh, that happens when using 2x the exact same column expression. I anticipated such issues when the column name is ambiguous (e.g. `AUTHOR.ID`, `BOOK.ID`) and we have integration tests for that, but not for when the exact same column is re-used.

Can you alias your column expressions as a workaround?

```java
  val nestedResults =
      dslContext
          .select(DSL.multiset(DSL.select(PROJECTS.ID, PROJECTS.ID).from(PROJECTS)))
          .fetch()
  println(nestedResults)
```

**Comment by sgrimm** [comment]
Ah, thanks. Using column aliases does indeed work around the problem.

**Comment by lukaseder** [comment]
It's not actually a `MULTISET` problem, but a bug in an underlying implementation of `Record.from(Object)`. This test also fails:

```java
@Test
public void testFromListDuplicateColumns() {
    Record2<Integer, Integer> r = create.newRecord(FIELD_ID1, FIELD_ID1);

    r.from(asList(1, 2));
    assertEquals(asList(1, 2), r.intoList());
}
```

The produced list is `[2, null]`

**Comment by lukaseder** [comment]
Yeah, the bug is in `DefaultRecordUnmapper.IterableUnmapper`, which uses name based field access, instead of index based access, which is both slower and ambiguous.

**Comment by lukaseder** [comment]
The same is true when passing an array

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.6 (https://github.com/jOOQ/jOOQ/issues/12710). Thanks again for your report!

**Comment by lukaseder** [comment]
This fix seems to have produced a regression when passing arrays to `Record.from(array, fields)`: https://github.com/jOOQ/jOOQ/issues/13040
